dali/kernels/imgproc/convolution/filter_gpu.cuh

```diff
@@ -0,0 +1,1094 @@
+ * memory. This is how different border modes (apart from BORDER_VALID) are handled.
+ * @{
+ */
+template <typename In, int axes, BoundaryType border>
+struct InLoaderBorderRemap {
+  template <BoundaryType border_>
+  using BorderTag = std::integral_constant<BoundaryType, border_>;
+
+  DALI_HOST_DEV DALI_FORCEINLINE int border_remap(int idx, int len,
+                                                  BorderTag<BoundaryType::REFLECT_101>) const {
+    assert(len > 0);
+    return boundary::idx_reflect_101(idx, len);
+  }
+
+  DALI_HOST_DEV DALI_FORCEINLINE int border_remap(int idx, int len,
+                                                  BorderTag<BoundaryType::REFLECT_1001>) const {
+    assert(len > 0);
+    return boundary::idx_reflect_1001(idx, len);
+  }
+
+  DALI_HOST_DEV DALI_FORCEINLINE int border_remap(int idx, int len,
+                                                  BorderTag<BoundaryType::CLAMP>) const {
+    assert(len > 0);
+    return boundary::idx_clamp(idx, 0, len);
+  }
+
+  DALI_HOST_DEV DALI_FORCEINLINE int border_remap(int idx, int len,
+                                                  BorderTag<BoundaryType::WRAP>) const {
+    assert(len > 0);
+    return boundary::idx_wrap(idx, len);
+  }
+
+  DALI_HOST_DEV DALI_FORCEINLINE int border_remap(int idx, const ShapeDesc<axes>& sample_shape,
+                                                  int axis) const {
+    return border_remap(idx, sample_shape.in_extents[axis], BorderTag<border>{});
+  }
+
+  DALI_HOST_DEV DALI_FORCEINLINE int border_remap_innermost(
+      int idx, const ShapeDesc<axes>& sample_shape) const {
+    int num_channels = sample_shape.num_channels;
+    if (idx < 0) {
+      // First, shift by 1 towards 0 (idx + 1), so that indices belonging to the same pixel
+      // translate to the same number pixel index when dividing by the channels stride
+      // (-6, -5, -4, -3, -2, -1) + 1 -> (-5, -4, -3, -2, -1, 0)
+      // (-5, -4, -3, -2, -1, 0) / 3 -> (-1, -1, -1, 0, 0, 0)
+      // Then shift back away from 0 (-1, -1, -1, 0, 0, 0) - 1 -> (-2, -2, -2, -1, -1, -1)
+      // Finally, with (num_channels - 1) we get the positive channels indecies
+      // (-2, -1, 0, -2, -1, 0) + 3 - 1 -> (0, 1, 2, 0, 1, 2)
+      int reflect_dim_idx = (idx + 1) / num_channels - 1;
+      int inner_dim_idx = (idx + 1) % num_channels + num_channels - 1;
+      return border_remap(reflect_dim_idx, sample_shape.width, BorderTag<border>{}) * num_channels +
+             inner_dim_idx;
+    }
+    if (idx >= sample_shape.in_extents.x) {
+      return border_remap(idx / num_channels, sample_shape.width, BorderTag<border>{}) *
+                 num_channels +
+             idx % num_channels;
+    }
+    return idx;
+  }
+
+  DALI_HOST_DEV DALI_FORCEINLINE In load(const In* __restrict__ in, const ivec<axes>& coords,
+                                         const ShapeDesc<axes>& sample_shape) const {
```

Comment:
A general thought. I see you are passing arguments by reference in those functions. While optimizing other kernels, we found that it generally works generally better to pass those by value. It'd worth trying to see if you can squeeze a better performance by doing that.

Reply:
That's interesting, let me check that out. My intuition was that constness + reference should perform no worse than passing by value.